File: push/src/main/java/com/xiaomi/push/service/MIPushNotificationCacheSupport.java
package com.xiaomi.push.service;

import android.content.Context;
import android.text.TextUtils;
import android.util.Pair;
import com.xiaomi.xmpush.thrift.PushMetaInfo;
import com.xiaomi.xmpush.thrift.XmPushActionContainer;
import java.util.LinkedList;

final class MIPushNotificationCacheSupport {
    private MIPushNotificationCacheSupport() {
    }

    static void clearNotification(Context context, String str, LinkedList<Pair<Integer, XmPushActionContainer>> linkedList) {
        clearNotification(context, str, -1, linkedList);
    }

    static void clearNotification(Context context, String str, int i, LinkedList<Pair<Integer, XmPushActionContainer>> linkedList) {
        NotificationManagerHelper notificationManagerHelperFrom = NotificationManagerHelper.from(context, str);
        int iHashCode = ((str.hashCode() / 10) * 10) + i;
        LinkedList<Pair<Integer, XmPushActionContainer>> linkedList2 = new LinkedList<>();
        if (i >= 0) {
            notificationManagerHelperFrom.cancel(iHashCode);
        }
        synchronized (linkedList) {
            for (Pair<Integer, XmPushActionContainer> pair : linkedList) {
                XmPushActionContainer xmPushActionContainer = pair.second;
                if (xmPushActionContainer != null) {
                    String targetPackage = MIPushNotificationHelper.getTargetPackage(xmPushActionContainer);
                    if (i >= 0) {
                        if (iHashCode == pair.first.intValue() && TextUtils.equals(targetPackage, str)) {
                            linkedList2.add(pair);
                        }
                    } else if (i == -1 && TextUtils.equals(targetPackage, str)) {
                        notificationManagerHelperFrom.cancel(pair.first.intValue());
                        linkedList2.add(pair);
                    }
                }
            }
            linkedList.removeAll(linkedList2);
        }
        uploadClearMessageData(context, linkedList2);
    }

    static void clearNotification(Context context, String str, String str2, String str3, LinkedList<Pair<Integer, XmPushActionContainer>> linkedList) {
        if (TextUtils.isEmpty(str2) && TextUtils.isEmpty(str3)) {
            return;
        }
        LinkedList<Pair<Integer, XmPushActionContainer>> linkedList2 = new LinkedList<>();
        synchronized (linkedList) {
            for (Pair<Integer, XmPushActionContainer> pair : linkedList) {
                XmPushActionContainer xmPushActionContainer = pair.second;
                if (xmPushActionContainer != null) {
                    String targetPackage = MIPushNotificationHelper.getTargetPackage(xmPushActionContainer);
                    PushMetaInfo metaInfo = xmPushActionContainer.getMetaInfo();
                    if (metaInfo != null && TextUtils.equals(targetPackage, str)) {
                        String title = metaInfo.getTitle();
                        String description = metaInfo.getDescription();
                        if (!TextUtils.isEmpty(title) && !TextUtils.isEmpty(description) && checkMatch(str2, title) && checkMatch(str3, description)) {
                            NotificationManagerHelper.from(context, str).cancel(pair.first.intValue());
                            linkedList2.add(pair);
                        }
                    }
                }
            }
            linkedList.removeAll(linkedList2);
        }
        uploadClearMessageData(context, linkedList2);
    }

    static void cacheNotification(Pair<Integer, XmPushActionContainer> pair, LinkedList<Pair<Integer, XmPushActionContainer>> linkedList) {
        synchronized (linkedList) {
            linkedList.add(pair);
            if (linkedList.size() > MIPushNotificationHelper.MAX_NOTIFY_ID_CACHE_SIZE) {
                linkedList.remove();
            }
        }
    }

    static void uploadClearMessageData(Context context, LinkedList<? extends Object> linkedList) {
        if (linkedList == null || linkedList.size() <= 0) {
            return;
        }
        TinyDataHelper.cacheTinyData(context, PushConstants.DOT_CATEGORY_CLEAR_NOTIFICATION, PushConstants.CLEAR_NOTIFICATION, linkedList.size(), "");
    }

    private static boolean checkMatch(String str, String str2) {
        return TextUtils.isEmpty(str) || str2.contains(str);
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/push/service/MIPushNotificationCacheSupport.kt
package com.xiaomi.push.service

import android.content.Context
import android.text.TextUtils
import android.util.Pair
import com.xiaomi.xmpush.thrift.PushMetaInfo
import com.xiaomi.xmpush.thrift.XmPushActionContainer
import java.util.LinkedList

object MIPushNotificationCacheSupport {
    @JvmStatic
    fun clearNotification(
        context: Context,
        packageName: String,
        notifications: LinkedList<Pair<Int, XmPushActionContainer>>,
    ) {
        clearNotification(context, packageName, -1, notifications)
    }

    @JvmStatic
    fun clearNotification(
        context: Context,
        packageName: String,
        notificationId: Int,
        notifications: LinkedList<Pair<Int, XmPushActionContainer>>,
    ) {
        val notificationManager = NotificationManagerHelper.from(context, packageName)
        val hashedNotificationId = (packageName.hashCode() / 10) * 10 + notificationId
        val matchedNotifications = LinkedList<Pair<Int, XmPushActionContainer>>()
        if (notificationId >= 0) {
            notificationManager.cancel(hashedNotificationId)
        }
        synchronized(notifications) {
            for (pair in notifications) {
                val container = pair.second ?: continue
                val targetPackage = MIPushNotificationHelper.getTargetPackage(container)
                if (notificationId >= 0) {
                    if (hashedNotificationId == pair.first && TextUtils.equals(targetPackage, packageName)) {
                        matchedNotifications.add(pair)
                    }
                } else if (notificationId == -1 && TextUtils.equals(targetPackage, packageName)) {
                    notificationManager.cancel(pair.first)
                    matchedNotifications.add(pair)
                }
            }
            notifications.removeAll(matchedNotifications)
        }
        uploadClearMessageData(context, matchedNotifications)
    }

    @JvmStatic
    fun clearNotification(
        context: Context,
        packageName: String,
        titleFilter: String,
        descriptionFilter: String,
        notifications: LinkedList<Pair<Int, XmPushActionContainer>>,
    ) {
        if (TextUtils.isEmpty(titleFilter) && TextUtils.isEmpty(descriptionFilter)) {
            return
        }
        val matchedNotifications = LinkedList<Pair<Int, XmPushActionContainer>>()
        synchronized(notifications) {
            for (pair in notifications) {
                val container = pair.second ?: continue
                val targetPackage = MIPushNotificationHelper.getTargetPackage(container)
                val metaInfo: PushMetaInfo = container.metaInfo ?: continue
                if (TextUtils.equals(targetPackage, packageName)) {
                    val title = metaInfo.title
                    val description = metaInfo.description
                    if (!TextUtils.isEmpty(title) &&
                        !TextUtils.isEmpty(description) &&
                        checkMatch(titleFilter, title) &&
                        checkMatch(descriptionFilter, description)
                    ) {
                        NotificationManagerHelper.from(context, packageName).cancel(pair.first)
                        matchedNotifications.add(pair)
                    }
                }
            }
            notifications.removeAll(matchedNotifications)
        }
        uploadClearMessageData(context, matchedNotifications)
    }

    @JvmStatic
    fun cacheNotification(
        notification: Pair<Int, XmPushActionContainer>,
        notifications: LinkedList<Pair<Int, XmPushActionContainer>>,
    ) {
        synchronized(notifications) {
            notifications.add(notification)
            if (notifications.size > MIPushNotificationHelper.MAX_NOTIFY_ID_CACHE_SIZE) {
                notifications.remove()
            }
        }
    }

    @JvmStatic
    fun uploadClearMessageData(context: Context, notifications: LinkedList<*>) {
        if (notifications.isNullOrEmpty()) {
            return
        }
        TinyDataHelper.cacheTinyData(
            context,
            PushConstants.DOT_CATEGORY_CLEAR_NOTIFICATION,
            PushConstants.CLEAR_NOTIFICATION,
            notifications.size.toLong(),
            "",
        )
    }

    private fun checkMatch(filter: String, value: String): Boolean {
        return TextUtils.isEmpty(filter) || value.contains(filter)
    }
}
